File: java/src/br/unifor/facilitador/aplicacao/AnalisadorIA.java
package br.unifor.facilitador.aplicacao;

import br.unifor.facilitador.dominio.*;
import br.unifor.facilitador.porta.*;
import java.util.ArrayList;
import java.util.List;

/**
 * UM analisador para TODAS as especialidades.
 * O que muda entre cardiologia e dermatologia e o prompt e a lista de exames
 * esperados, e isso vem do banco como DADO. Nova especialidade = novo registro.
 */
public class AnalisadorIA implements AnalisadorClinico {

    private final ConfigEspecialidadeRepository configs;
    private final ProvedorIA provedor;

    public AnalisadorIA(ConfigEspecialidadeRepository configs, ProvedorIA provedor) {
        this.configs = configs;
        this.provedor = provedor;
    }

    @Override
    public Sugestao analisar(Consulta c, List<Documento> docs) {
        ConfiguracaoEspecialidade cfg = configs.para(c.getMedico().getEspecialidade());

        List<TipoDocumento> faltantes = new ArrayList<>();
        for (TipoDocumento esperado : cfg.getDocsEsperados()) {
            boolean achou = docs.stream().anyMatch(d -> d.getTipo() == esperado);
            if (!achou) {
                faltantes.add(esperado);
            }
        }

        String prompt = montarPrompt(cfg, docs);
        String resposta = provedor.completar(prompt);

        return new Sugestao(
                "Triagem de " + cfg.getEspecialidade() + " com " + docs.size() + " documento(s).",
                resposta,
                faltantes);
    }

    private String montarPrompt(ConfiguracaoEspecialidade cfg, List<Documento> docs) {
        StringBuilder sb = new StringBuilder(cfg.getPromptBase());
        sb.append("\nDocumentos recebidos:");
        for (Documento d : docs) {
            sb.append("\n- ").append(d.getTipo());
        }
        return sb.toString();
    }
}

File: javascript/src/aplicacao/AnalisadorIA.js
import { AnalisadorClinico } from '../porta/AnalisadorClinico.js';
import { Sugestao } from '../dominio/Sugestao.js';

/**
 * UM analisador para TODAS as especialidades.
 * O que muda entre cardiologia e dermatologia e o prompt e a lista de exames
 * esperados, e isso vem do banco como DADO. Nova especialidade = novo registro.
 */
export class AnalisadorIA extends AnalisadorClinico {
  #configs;
  #provedor;

  constructor(configs, provedor) {
    super();
    this.#configs = configs;
    this.#provedor = provedor;
  }

  analisar(consulta, documentos) {
    const cfg = this.#configs.para(consulta.medico.especialidade);

    const faltantes = cfg.docsEsperados.filter(
      (esperado) => !documentos.some((d) => d.tipo === esperado),
    );

    const prompt = this.#montarPrompt(cfg, documentos);
    const resposta = this.#provedor.completar(prompt);

    return new Sugestao(
      `Triagem de ${cfg.especialidade} com ${documentos.length} documento(s).`,
      resposta,
      faltantes,
    );
  }

  #montarPrompt(cfg, documentos) {
    const lista = documentos.map((d) => `- ${d.tipo}`).join('\n');
    return `${cfg.promptBase}\nDocumentos recebidos:\n${lista}`;
  }
}
